jsonrpc/websocket.go

```diff
@@ -56,20 +57,17 @@ func (ws *Websocket) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	wsc := newWebsocketConn(r.Context(), conn, ws.connParams)
 
-	ctx := context.WithValue(wsc.ctx, ConnKey{}, wsc)
 	for {
-		var msg []byte
-		_, msg, err = wsc.conn.Read(ctx)
+		_, wsc.r, err = wsc.conn.Reader(wsc.ctx)
 		if err != nil {
 			break
 		}
-		var resp []byte
-		resp, err = ws.rpc.Handle(ctx, msg)
-		if err != nil {
+		ws.listener.OnNewRequest("any")
+		if err = ws.rpc.HandleReadWriter(wsc.ctx, wsc); err != nil {
 			break
 		}
-		ws.listener.OnNewRequest("any")
-		if _, err = wsc.Write(resp); err != nil {
+		// From websocket docs: "Read to EOF otherwise connection will hang."
+		if _, err = io.ReadAll(wsc.r); err != nil {
```

Comment:
Do you think it is worth it to use `io.Discard` https://pkg.go.dev/io#Writer to avoid the allocation?

Reply:
How would we use `io.Discard` to flush the reader?

Reply:
https://pkg.go.dev/io#Copy

Reply:
`io.Copy` will allocate regardless https://github.com/golang/go/blob/2744155d369ca838be57d1eba90c3c6bfc4a3b30/src/io/io.go#L426

Reply:
Oh missed this https://github.com/golang/go/blob/2744155d369ca838be57d1eba90c3c6bfc4a3b30/src/io/io.go#L634-L635

I'll use `io.Discard` :+1: